php-checks/src/main/java/org/sonar/php/checks/ParameterReassignedToCheck.java

```diff
@@ -0,0 +1,158 @@
+    ControlFlowGraph cfg = ControlFlowGraph.build(tree, context());
+    if (cfg == null) {
+      return new HashSet<>();
+    }
+
+    LiveVariablesAnalysis analysis = LiveVariablesAnalysis.analyze(cfg, context().symbolTable());
+    Set<Symbol> live = analysis.getLiveVariables(cfg.start()).getIn();
+    for (ParameterTree parameterTree : tree.parameters().parameters()) {
+      if (parameterTree.referenceToken() == null) {
+        Symbol symbol = context().symbolTable().getSymbol(parameterTree.variableIdentifier());
+        if (!live.contains(symbol)) {
+          investigatedParameters.add(symbol);
+        }
+      }
+    }
+    return live;
+  }
+
+  private void clearFixFunctionTree(FunctionTree tree, Set<Symbol> live) {
+    for (ParameterTree parameterTree : tree.parameters().parameters()) {
+      Symbol symbol = context().symbolTable().getSymbol(parameterTree.variableIdentifier());
+      if (!live.contains(symbol)) {
+        investigatedParameters.remove(symbol);
+      }
+    }
+  }
+
+  @Override
+  public void visitForEachStatement(ForEachStatementTree tree) {
+    ControlFlowGraph cfg = ControlFlowGraph.build(tree, context());
+    if (cfg == null) {
+      return;
+    }
+
+    Symbol symbol = context().symbolTable().getSymbol(tree.value());
+    LiveVariablesAnalysis analysis = LiveVariablesAnalysis.analyze(cfg, context().symbolTable());
+
+    Set<Symbol> live = analysis.getLiveVariables(cfg.start()).getIn();
+    boolean liveVar = live.contains(symbol);
+
+    if (!liveVar) {
+      investigatedParameters.add(symbol);
+    }
+
+    super.visitForEachStatement(tree);
+
+    if (!liveVar) {
+      investigatedParameters.remove(symbol);
+    }
+  }
+
+  @Override
+  public void visitCatchBlock(CatchBlockTree tree) {
+    ControlFlowGraph cfg = ControlFlowGraph.build(tree, context());
+    if (cfg == null) {
+      return;
+    }
+
+    Symbol symbol = context().symbolTable().getSymbol(tree.variable());
+    LiveVariablesAnalysis analysis = LiveVariablesAnalysis.analyze(cfg, context().symbolTable());
+
+    Set<Symbol> live = analysis.getLiveVariables(cfg.start()).getIn();
+    boolean liveVar = live.contains(symbol);
+
+    if (!liveVar) {
+      investigatedParameters.add(symbol);
+    }
+
+    super.visitCatchBlock(tree);
+
+    if (!liveVar) {
+      investigatedParameters.remove(symbol);
+    }
+  }
+
+  @Override
+  public void visitAssignmentExpression(AssignmentExpressionTree tree) {
+    if (tree.variable().is(Tree.Kind.VARIABLE_IDENTIFIER)) {
+      VariableIdentifierTree identifier = (VariableIdentifierTree) tree.variable();
+      Symbol reference = context().symbolTable().getSymbol(identifier);
+      if (reference != null && (reference.is(Symbol.Kind.PARAMETER) || reference.is(Symbol.Kind.VARIABLE)) && investigatedParameters.contains(reference)) {
+        context().newIssue(this, tree, String.format(MESSAGE, reference.name())).secondary(reference.declaration(), SECONDARY_MESSAGE);
```

Comment:
Don't you want to add an assertion about the message in one of your test cases? Something like:
```
... // Noncompliant {{Introduce a new variable instead of reusing the parameter \"foo\".}}
```